lib/email/receiver.rb

```diff
@@ -745,23 +745,23 @@ def process_destination(destination, user, body, elided)
 
     def create_group_post(group, user, body, elided)
       message_ids = Email::Receiver.extract_reply_message_ids(@mail, max_message_id_count: 5)
-      post_ids = []
 
+      # incoming emails with matching message ids, and then cross references
+      # these with any email addresses for the user vs to/from/cc of the
+      # incoming emails. in effect, any incoming email record for these
+      # message ids where the user is involved in any way will be returned
       incoming_emails = IncomingEmail.where(message_id: message_ids)
       if !group.allow_unknown_sender_topic_replies
         incoming_emails = incoming_emails.addressed_to_user(user)
       end
-
-      incoming_emails = incoming_emails.pluck(:post_id, :from_address, :to_addresses, :cc_addresses)
-
-      incoming_emails.each do |post_id, from_address, to_addresses, cc_addresses|
-        if group.allow_unknown_sender_topic_replies
-          post_ids << post_id
-        else
-          post_ids << post_id if contains_email_address_of_user?(from_address, user) ||
-            contains_email_address_of_user?(to_addresses, user) ||
-            contains_email_address_of_user?(cc_addresses, user)
-        end
+      post_ids = incoming_emails.pluck(:post_id) || []
+
+      # if the user is directly replying to an email send to them from discourse,
+      # there will be a corresponding EmailLog record, so we can use that as the
+      # reply post if it exists
+      if discourse_generated_message_id?(mail.in_reply_to)
+        email_log = EmailLog.where(message_id: mail.in_reply_to).addressed_to_user(user).last
```

Comment:
Minor perf concern here. Is it possible to just pluck the `post_id` so we don't have to load the entire ActiveRecord object?